plugins/builds/index.js

```diff
@@ -293,83 +410,329 @@ exports.register = (server, options, next) => {
+                    [jName]: null
                 };
+            });
+            // need to deepmerge because it's possible same event has multiple builds
+            // const combined = deepmerge(currentJobParentBuilds, joinParentBuilds);
+
+            // override currentBuild in the joinParentBuilds
+            const currentBuildInfo = {
+                [pipelineId]: {
+                    eventId: build.eventId,
+                    [currentJobName]: build.id
+                }
+            };
+            // const parentBuilds = deepmerge(combined, currentBuildInfo);
+            const parentBuilds = deepmerge.all(
+                [joinParentBuilds, currentJobParentBuilds, currentBuildInfo]);
+
+            /* CONSTRUCT BUILD CONFIG TO CREATE */
+            const internalBuildConfig = {
+                jobFactory,
+                buildFactory,
+                eventFactory,
+                pipelineId,
+                jobName: nextJobName,
+                username,
+                scmContext,
+                build, // this is the parentBuild for the next build
+                baseBranch: event.baseBranch || null,
+                parentBuilds
+            };
+            const externalBuildConfig = {
+                pipelineFactory,
+                eventFactory,
+                externalPipelineId,
+                externalJobName,
+                parentBuildId: build.id,
+                parentBuilds,
+                causeMessage: `Triggered by sd@${pipelineId}:${currentJobName}`
+            };
+
+            /* CREATE AND START NEXT BUILD IF ALL INTO 3 SCENARIOS
+             * 1. No join
+             * 2. ([~D,B,C]->A) currentJob=D, nextJob=A, joinList(A)=[B,C]
+             *    joinList doesn't include D, so start A
+             * 3. ([~D,B,C]-> sd@123:A) currentJob=D, nextJob=sd@123:A, joinList(A)=[sd@111:B,sd@111:C]
+             *    joinList doesn't include sd@111:D, so start A
+             */
+            // current job can be "external" in nextJob's perspective
+            const currentJobNotInJoinList = !joinListNames.includes(currentJobName) &&
+                !joinListNames.includes(`sd@${pipelineId}:${currentJobName}`);
+
+            if (joinList.length === 0 || currentJobNotInJoinList) {
+                if (!isExternal) {
+                    return createInternalBuild(internalBuildConfig);
+                }
+
+                return createExternalBuild(externalBuildConfig);
+            }
+
+            let nextBuild;
+
+            /* CHECK WHETHER NEXT BUILD EXISTS */
+            // if next build is external, check the latest build
+            if (isExternal) {
+                const p = await pipelineFactory.get(externalPipelineId);
+                const jobArray = await p.getJobs({ params: { name: externalJobName } });
+                const j = await jobFactory.get(jobArray[0].id);
+
+                nextBuild = await j.getLatestBuild({ status: 'CREATED' });
```

Comment:
per @jithin1987  suggestion, we should store a new field maybe `upstreamEvent`, to trace back to previous event. So that we can get the actual build instead of just fetching the latest build here